### DISCUSSION TITLE
Type safety in property values

### DISCUSSION THREAD

**Discussion by TowardsDeath** [chakra-ui/panda#1232] [Q&A]
I‘m trying Panda CSS to see if it‘s something we could use in our design system, instead of Vanilla Extract. I‘m having trouble getting type-checking in CSS values to work.

From the [documentation](https://panda-css.com/docs/concepts/writing-styles#type-safety):

> Panda is built with TypeScript and provides type safety for all style properties and shorthands.

Does this mean only the properties are type-checked, and the values aren‘t?

This fails like I would expect:
```
css({
  positio: 'absolut'
})
```
```
Type error: Argument of type '{ positio: string; }' is not assignable to parameter of type 'SystemStyleObject'.
  Object literal may only specify known properties, but 'positio' does not exist in type 'SystemStyleObject'. Did you mean to write 'position'?

> 11 |   positio: 'absolut',
```
This passes, even though the value is invalid:

```
css({
  position: 'absolut'
})
```
Am I doing something wrong, or is this how Panda CSS behaves?

**Comment by anubra266** [comment]
@TowardsDeath css allows arbitrary values naturally. For example, `height: 4px` we can't have typescript restrict all these values right.
So if you want to prevent arbitrary values. You should use [strictTokens](https://panda-css.com/docs/references/config#stricttokens).

**Reply by TowardsDeath** [reply]
@anubra266 thanks for your reply.

> css allows arbitrary values naturally. For example, height: 4px we can't have typescript restrict all these values right.

No, of course, but lots of CSS properties have a fixed set of possible values, in which case type checking would be a helpful addition. They come with CSSType, which Panda already uses, if I'm not mistaken.

Vanilla Extract does catch these errors for these properties, and still allows other values for other properties, like `height: 4px`. <sup>1</sup>

```
css/index.css.ts:10:3 - error TS2820: Type '"absolut"' is not assignable to type 'CSSVarFunction | Position | (CSSVarFunction | Position | undefined)[] | undefined'. Did you mean '"absolute"'?
```

> So if you want to prevent arbitrary values. You should use [strictTokens](https://panda-css.com/docs/references/config#stricttokens).

For some reason, this doesn't do anything in my project (default setup with Next.js app router and Panda). If it works, does this throw an error when someone makes a mistake in the value, e.g. `position: absolut`?

---

<sup>1</sup> This is not to discredit all of the lovely work and effort that has been put into Panda CSS, nor am I implying that Vanilla Extract is better. I will still be using Panda for my personal website. It's just that type safety is more important at work, so I'd like to make an informed decision about the future of our component library.

**Reply by tombstonedguy** [reply]
it seems that the [strictTokens](https://panda-css.com/docs/references/config#stricttokens)-Option doesn´t work anymore from version 0.11 on.
see [Issue](https://github.com/chakra-ui/panda/issues/815)

**Comment by anubra266** [comment]
@TowardsDeath is this still an issue for you?

**Reply by TowardsDeath** [reply]
Yes, `strictTokens: true` does not catch `position: 'absolut'`. Would you consider expanding the type definitions by using the types from CSSType, instead of `PropertyValue<'position'>`? In the case of `position`:

```
export type Position = Globals | "-webkit-sticky" | "absolute" | "fixed" | "relative" | "static" | "sticky";
```
Source: https://github.com/frenic/csstype/blob/master/index.d.ts#L19411

**Reply by anubra266** [reply]
cc @astahmer

**Reply by astahmer** [reply]
hey, late to the party but this seems fixed
<img width="802" alt="Screenshot 2023-10-02 at 11 02 44" src="https://github.com/chakra-ui/panda/assets/47224540/ed62ca74-3ebe-430e-a416-cfd4d3f452f6">

**Reply by TowardsDeath** [reply]
Yes, thanks! Just upgraded to 0.15.5 and can confirm the type checks are more strict now.
